src/super_gradients/training/datasets/samplers/class_balanced_sampler.py

```diff
@@ -0,0 +1,151 @@
+        category_repeat = {
+            cat_id: cat_repeat
+            for cat_id, cat_repeat in enumerate(_default_oversample_heuristic(class_frequencies, oversample_threshold, oversample_aggressiveness))
+        }  # dict for ease of debugging
+
+        # 3. For each image I, compute the image-level repeat factor: r(I) = max_{c in I} r(c)
+        repeat_factors = list()
+        categories = np.arange(class_information.shape[1])
+        for sample_cat_freq in class_information:
+            cat_ids = categories[sample_cat_freq != 0]
+            if len(cat_ids) == 0:  # in case image doesn't have annotations, we will not over-sample nor ignore it
+                repeat_factors.append(1.0)
+            else:
+                repeat_factors.append(max({category_repeat[cat_id] for cat_id in cat_ids}))
+
+        return repeat_factors  # len = dataset_length
+
+    @staticmethod
+    def precompute_sample_repeat_factors(
+        output_path: str,
+        class_information_provider: HasClassesInformation,
+        oversample_threshold: Optional[float] = None,
+    ):
+        repeat_factors: List[float] = ClassBalancer.get_sample_repeat_factors(
+            class_information_provider=class_information_provider,
+            oversample_threshold=oversample_threshold,
+        )
+
+        str_repeat_factors = [np.format_float_positional(rf, trim="0", precision=4) for rf in repeat_factors]
+
+        with open(output_path, "w") as f:
+            json.dump(str_repeat_factors, f)
+
+    @staticmethod
+    def from_precomputed_sample_repeat_factors(precomputed_path: str) -> List[float]:
+        """
+        Loads the repeat factors from a precomputed file.
+        """
+        if not os.path.exists(precomputed_path):
+            raise FileNotFoundError(f"`{precomputed_path}` does not exist.")
+
+        with open(precomputed_path, "r") as f:
+            loaded = json.load(f)
+
+        return list(map(lambda x: float(x), loaded))
+
+
+@register_sampler(Samplers.CLASS_BALANCED)
+class ClassBalancedSampler(WeightedRandomSampler):
+    def __init__(
+        self,
+        dataset: Optional[HasClassesInformation] = None,
+        precomputed_factors_file: Optional[str] = None,
+        oversample_threshold: Optional[float] = None,
+        oversample_aggressiveness: float = 0.5,
+        generator=None,
+    ) -> None:
+        """
+        Wrap WeightedRandomSampler with weights that are computed from the class frequencies of the dataset.
+        """
+
+        if dataset is None and precomputed_factors_file is None:
+            raise ValueError("`dataset` and `precomputed_factors` cannot be both None.")
+
+        if dataset is not None and precomputed_factors_file is not None:
+            # this logic is to simplify `_instantiate_sampler` method.
+            warnings.warn("Both `dataset` and `precomputed_factors_file` are passed. `dataset` WILL BE IGNORED!")
```

Comment:
Because dataset is injected by `_instantiate_sampler`, there's a problem when `precomputed_factors_file` is passed, because then both appear.
We can either add logic to `_instantiate_sampler` (I don't think it should know the internals), or to prioritize `precomputed_factors_file` like implemented here.